File: Atividades/2024-03-13/Arvores/arvores_binarias_atividade/src/main/java/dev/mandevilla/Services/ArvoreBinariaService.java
package dev.mandevilla.Services;

import java.util.List;

import dev.mandevilla.Models.NoModel;

public class ArvoreBinariaService {
    protected NoModel raiz;

    public ArvoreBinariaService() {
        raiz = null;
    }

    public NoModel obterRaiz() {
        return raiz;
    }

    public void inserir(int valorRaiz) {
        if (raiz == null) {
            raiz = new NoModel(valorRaiz);
        }
        else {
            inserir(valorRaiz, raiz);
        }        
    }

    public List<Integer> obterPreOrdem(NoModel no, List<Integer> lista) {
        if (no != null) {
            lista.add(no.getValor());
            lista = obterPreOrdem(no.getEsquerdo(), lista);
            lista = obterPreOrdem(no.getDireito(), lista);
        }

        return lista;
    }

    public List<Integer> obterEmOrdem(NoModel no, List<Integer> lista) {
        if (no != null) {
            lista = obterEmOrdem(no.getEsquerdo(), lista);
            lista.add(no.getValor());
            lista = obterEmOrdem(no.getDireito(), lista);
        }

        return lista;
    }

    public List<Integer> obterPosOrdem(NoModel no, List<Integer> lista) {
        if (no != null) {
            lista = obterPosOrdem(no.getEsquerdo(), lista);
            lista = obterPosOrdem(no.getDireito(), lista);
            lista.add(no.getValor());
        }

        return lista;
    }

    public void inserir(List<Integer> valores) {
        for (Integer valor : valores) {
            inserir(valor);
        }
    }

    public NoModel buscar(int valor) {
        if (raiz == null) {
            return null;
        }
        return buscar(valor, raiz);
    }

    public boolean isFolha(NoModel no) {
        return no.getEsquerdo() == null && no.getDireito() == null;
    }

    // Métodos privados

    private void inserir(int valor, NoModel noAtual) {
        if (valor <= noAtual.getValor()) {
            if (noAtual.getEsquerdo() == null) {
                noAtual.setEsquerdo(valor);
            }
            else {
                inserir(valor, noAtual.getEsquerdo());
            }
        }
        else {
            if (noAtual.getDireito() == null) {
                noAtual.setDireito(valor);
            }
            else {
                inserir(valor, noAtual.getDireito());
            }
        }
    }

    private NoModel buscar(int valor, NoModel noAtual) {
        if (valor == noAtual.getValor()) {
            return noAtual;
        }
        else if (valor <= noAtual.getValor()) {
            if (noAtual.getEsquerdo() == null) {
                return null;
            }
            else {
                return buscar(valor, noAtual.getEsquerdo());
            }
        }
        else {
            if (noAtual.getDireito() == null) {
                return null;
            }
            else {
                return buscar(valor, noAtual.getDireito());
            }
        }
    }
} 

File: Atividades/2024-03-27/Arvores/arvores_binarias_simulado/lib/src/shared/services/arvore_binaria/arvore_binaria_service.dart
import 'package:arvores_binarias_simulado/src/shared/models/no/no_model.dart';

/// Classe de serviço, para operações com Árvores binárias.
class ArvoreBinariaService {
  /// Nó raiz da Árvore binária.
  NoModel? _noRaiz;

  /// Retorna o nó raiz da Árvore binária.
  NoModel get noRaiz {
    if (_noRaiz is NoModel) {
      return _noRaiz!;
    }

    throw Exception('A raiz da árvore binária é nula.');
  }

  /// Insere um valor na Árvore binária.
  void inserir(int valor) {
    if (_noRaiz == null) {
      _noRaiz = NoModel(valor: valor);
    } else {
      _inserir(valor, _noRaiz!);
    }
  }

  /// Insere uma lista de valores na Árvore binária.
  void inserirLista(List<int> valores) {
    for (var valor in valores) {
      inserir(valor);
    }
  }

  /// Busca um nó na Árvore binária, baseado num valor informado.
  NoModel? buscar(int valor) {
    if (_noRaiz == null) {
      return null;
    }

    return _buscar(valor, _noRaiz!);
  }

  /// Retorna uma lista de valores inteiros, referentes aos nós da Árvore binária
  /// dispostos em **pré-ordem**.
  ///
  /// 1. Raiz;
  ///
  /// 2. Esquerda;
  ///
  /// 3. Direita;
  List<int> obterPreOrdem({
    required NoModel? no,
    required List<int> retorno,
  }) {
    if (no is NoModel) {
      retorno.add(no.valor);
      retorno = obterPreOrdem(no: no.esquerdo, retorno: retorno);
      retorno = obterPreOrdem(no: no.direito, retorno: retorno);
    }

    return retorno;
  }

  /// Retorna uma lista de valores inteiros, referentes aos nós da Árvore binária
  /// dispostos **em-ordem**.
  ///
  /// 1. Esquerda;
  ///
  /// 2. Raiz;
  ///
  /// 3. Direita;
  List<int> obterEmOrdem({
    required NoModel? no,
    required List<int> retorno,
  }) {
    if (no != null) {
      retorno = obterEmOrdem(no: no.esquerdo, retorno: retorno);
      retorno.add(no.valor);
      retorno = obterEmOrdem(no: no.direito, retorno: retorno);
    }

    return retorno;
  }

  /// Retorna uma lista de valores inteiros, referentes aos nós da Árvore binária
  /// dispostos em **pós-ordem**.
  ///
  /// 1. Esquerda;
  ///
  /// 2. Direita;
  ///
  /// 3. Raiz;
  List<int> obterPosOrdem({
    required NoModel? no,
    required List<int> retorno,
  }) {
    if (no is NoModel) {
      retorno = obterPosOrdem(no: no.esquerdo, retorno: retorno);
      retorno = obterPosOrdem(no: no.direito, retorno: retorno);
      retorno.add(no.valor);
    }

    return retorno;
  }

  // Métodos privados

  /// Insere um valor na Árvore binária, respeitando as regras da estrutura de dados.
  void _inserir(int valor, NoModel noAtual) {
    if (valor <= noAtual.valor) {
      if (noAtual.esquerdo is NoModel) {
        _inserir(valor, noAtual.esquerdo!);
      } else {
        noAtual.setEsquerdo(valor);
      }
    } else {
      if (noAtual.direito is NoModel) {
        _inserir(valor, noAtual.direito!);
      } else {
        noAtual.setDireito(valor);
      }
    }
  }

  /// Busca a primeira ocorrência de um valor informado.
  NoModel? _buscar(int valor, NoModel noAtual) {
    if (valor == noAtual.valor) {
      return noAtual;
    } else if (valor <= noAtual.valor) {
      if (noAtual.esquerdo is NoModel) {
        return _buscar(valor, noAtual.esquerdo!);
      } else {
        return null;
      }
    } else {
      if (noAtual.direito is NoModel) {
        return _buscar(valor, noAtual.direito!);
      } else {
        return null;
      }
    }
  }
}
